core/src/readwriter.rs

```diff
@@ -302,6 +304,105 @@ impl DataFrameWriter {
     }
 }
 
+pub struct DataFrameWriterV2 {
+    dataframe: DataFrame,
+    table: String,
+    provider: Option<String>,
+    options: HashMap<String, String>,
+    properties: HashMap<String, String>,
+    partitioning: Vec<Expression>,
+    overwrite_condition: Option<Expression>,
+}
+
+impl DataFrameWriterV2 {
+    pub fn new(dataframe: DataFrame, table: &str) -> Self {
+        Self {
+            dataframe,
+            table: table.to_string(),
+            provider: None,
+            options: HashMap::new(),
+            properties: HashMap::new(),
+            partitioning: vec![],
+            overwrite_condition: None,
+        }
+    }
+
+    pub fn using(mut self, provider: &str) -> Self {
+        self.provider.replace(provider.to_string());
+        self
+    }
+
+    pub fn option(mut self, key: &str, value: &str) -> Self {
+        self.options.insert(key.to_string(), value.to_string());
+        self
+    }
+
+    pub fn options(mut self, provider: HashMap<String, String>) -> Self {
+        self.options.extend(provider);
+        self
+    }
+
+    #[allow(non_snake_case)]
+    pub fn tableProperty(mut self, property: &str, value: &str) -> Self {
+        self.properties
+            .insert(property.to_string(), value.to_string());
+        self
+    }
+
+    #[allow(non_snake_case)]
+    pub fn partitionBy(mut self, columns: Vec<Expression>) -> Self {
```

Comment:
Sorry, I missed this. For this method, lets use the custom trait `ToVecExpr` for this input type. 

So the method would look like this. 

```rust
    pub fn partitionBy<T: ToVecExpr>(mut self, columns: T) -> Self {
        self.partitioning = columns.to_vec_expr();
        self
    }
```

This will give the user the ability to have the input be different variations of 
`partitionBy("id")`, `partitionBy(col("id"))`, `partitionBy(vec![F::year(), F::month()])`,`partitionBy([F::year(), F::month()])`, etc.